### PR TITLE
Add vector read and write support to iterators

### PR DESCRIPTION
**Pull request by simba-git**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: provider/iterators.go [hunk 1]
```diff
@@ -76,6 +84,40 @@ func (p *parquetIterator) Next() bool {
...
 }
 
+// parseFloatVec parses a generic float array that is received via a parquet file. It shows up in the form:
+// map[list:[map[element: 1] map[element:2] map[element:3]]]
+func (p *parquetIterator) parseFloatVec(val map[string]interface{}) ([]float32, error) {
+	list, ok := val["list"]
+	if !ok {
+		return nil, fferr.NewDataTypeNotFoundErrorf(val, "expected to find field 'list' when parsing float vector")
+	}
+	// To iterate over the list and create a we need to cast it to []interface{}
```
Comment (sdreyer) [inline provider/iterators.go:94]: create a what

File: provider/offline.go [hunk 2]
```diff
@@ -551,28 +560,34 @@ func (schema *TableSchema) ToParquetRecords(records []GenericRecord) []any {
...
+					reflectValue = ogVal.Addr()
+				} else {
+					reflectValue = reflect.New(reflect.TypeOf(v))
+					reflectValue.Elem().Set(ogVal)
+				}
+			// Lists and timestamps
 			default:
-				parquetRecord.Elem().FieldByName(colName).Set(reflect.ValueOf(value))
+				reflectValue = reflect.ValueOf(v)
+				fmt.Printf("THERE %v %v\n", v, value)
```
Comment (sdreyer) [inline provider/offline.go:579]: prints

File: provider/offline.go [hunk 3]
```diff
@@ -551,28 +560,34 @@ func (schema *TableSchema) ToParquetRecords(records []GenericRecord) []any {
...
+				fmt.Printf("THERE %s\n", reflectValue.Type().String())
+				fmt.Printf("THERE %v\n", reflectValue.Interface())
+			}
+			if !reflectValue.Type().AssignableTo(parquetField.Type()) {
+				return nil, fferr.NewInternalErrorf("Writing Invalid Type to Parquet Record. Found %s expected %s", reflectValue.Type().String(), parquetField.Type().String())
```
Comment (epps) [inline provider/offline.go:584]: Repeating @anthonylasso's nit: lower case for error messages.

**Review by epps — APPROVED** [review]
🧑‍🚒

File: fferr/data.go [hunk 4]
```diff
@@ -52,18 +52,22 @@ type DatasetAlreadyExistsError struct {
...
 }
 
-func NewDataTypeNotFoundError(valueType string, err error) *DataTypeNotFoundError {
+func NewDataTypeNotFoundError(value any, err error) *DataTypeNotFoundError {
 	if err == nil {
 		err = fmt.Errorf("datatype not found")
 	}
 	baseError := newBaseError(err, DATATYPE_NOT_FOUND, codes.NotFound)
-	baseError.AddDetail("value_type", valueType)
+	baseError.AddDetail("value_and_type", fmt.Sprintf("%#v %T", value, value))
```
Comment (aolfat) [inline fferr/data.go:60]: if you wanted to, you could do baseError.AddDetails("value", fmt.Sprintf("%#v"), "type", "fmt.Sprintf("%T"))

File: provider/offline.go [hunk 5]
```diff
@@ -551,28 +560,34 @@ func (schema *TableSchema) ToParquetRecords(records []GenericRecord) []any {
...
-			case bool:
-				parquetRecord.Elem().FieldByName(colName).Set(reflect.ValueOf(&v))
+			case int, int32, int64, float32, float64, string, bool:
+				// Pointer types are used for all the scalar types to ensure they
+				// can be nullable in the parquet file.
+				ogVal := reflect.ValueOf(v)
+				if ogVal.CanAddr() {
+					reflectValue = ogVal.Addr()
+				} else {
+					reflectValue = reflect.New(reflect.TypeOf(v))
```
Comment (aolfat) [inline provider/offline.go:573]: i wonder if it makes sense to ever cache this? Don't know how inefficient it is in go but when doing someting similar like this in python for numpy it ended up being a bit of a bottlneck

**Pull request merged**